contracts/src/Assets.sol

```diff
@@ -44,54 +36,99 @@ library Assets {
 
         if (destinationChain == assetHubParaID) {
             if (destinationAddress.isAddress32()) {
-                payload = SubstrateTypes.SendTokenToAssetHubAddress32(token, destinationAddress.asAddress32(), amount);
+                ticket.payload = SubstrateTypes.SendTokenToAssetHubAddress32(
+                    token, destinationAddress.asAddress32(), $.assetHubReserveTransferFee, amount
+                );
             } else {
+                // AssetHub does not support 20-byte account IDs
                 revert Unsupported();
             }
         } else {
             if (destinationAddress.isAddress32()) {
-                payload = SubstrateTypes.SendTokenToAddress32(
-                    token, destinationChain, destinationAddress.asAddress32(), amount
+                ticket.payload = SubstrateTypes.SendTokenToAddress32(
+                    token,
+                    destinationChain,
+                    destinationAddress.asAddress32(),
+                    $.assetHubReserveTransferFee,
+                    destinationChainFee,
+                    amount
                 );
             } else if (destinationAddress.isAddress20()) {
-                payload = SubstrateTypes.SendTokenToAddress20(
-                    token, destinationChain, destinationAddress.asAddress20(), amount
+                ticket.payload = SubstrateTypes.SendTokenToAddress20(
+                    token,
+                    destinationChain,
+                    destinationAddress.asAddress20(),
+                    $.assetHubReserveTransferFee,
+                    destinationChainFee,
+                    amount
                 );
             } else {
                 revert Unsupported();
             }
         }
-        extraFee = $.sendTokenFee;
-
         emit IGateway.TokenSent(sender, token, destinationChain, destinationAddress, amount);
     }
 
-    /// @dev Enqueues a create native token message to substrate.
-    /// @param token The ERC20 token address.
-    function registerToken(address token) external returns (bytes memory payload, uint256 extraFee) {
+    function registerTokenCosts() external view returns (Costs memory costs) {
+        return _registerTokenCosts();
+    }
+
+    function _registerTokenCosts() internal view returns (Costs memory costs) {
         AssetsStorage.Layout storage $ = AssetsStorage.layout();
+        costs.foreign = $.assetHubCreateAssetFee;
+        costs.native = $.registerTokenFee;
+    }
 
+    /// @dev Enqueues a create native token message to substrate.
+    /// @param token The ERC20 token address.
+    function registerToken(address token) external returns (Ticket memory ticket) {
         if (!token.isContract()) {
             revert InvalidToken();
         }
 
-        payload = SubstrateTypes.RegisterToken(token);
-        extraFee = $.registerTokenFee;
+        ticket.costs = _registerTokenCosts();
+        ticket.payload = SubstrateTypes.RegisterToken(token);
```

Comment:
I realize that we should include `$.assetHubCreateAssetFee` in the message payload to BridgeHub, to be consistent with sendToken

```suggestion
        ticket.payload = SubstrateTypes.RegisterToken(token, $.assetHubCreateAssetFee);
```

Reply:
Seems we also need to have a `$.assetHubCreateAssetDeposit` to match the XCM conversion code in

https://github.com/Snowfork/snowbridge/blob/510880c717bf34941c75aff2ed4c63b272964fa3/parachain/primitives/router/src/inbound/mod.rs#L165

Reply:
Addressed in https://github.com/Snowfork/snowbridge/pull/1015/commits/edf418d7890a2d8d40e816cff1c0923e64840772

> currently assetHubCreateAssetFee is fee + deposit

Actually it's for different purposes, `assetHubCreateAssetFee` is only paying for xcm execution while  `CreateAssetDeposit` is for the required deposit(0.1 DOT).
https://github.com/Snowfork/snowbridge/blob/510880c717bf34941c75aff2ed4c63b272964fa3/parachain/primitives/router/src/inbound/mod.rs#L177-L180

Ideally the `registerTokenFee`(in Ether) should match the `CreateAssetDeposit`(in DOT), but  I would suggest we keep it seperated, we can charge more than the required deposit to avoid spamming.

Reply:
sounds fine 👍